### DISCUSSION TITLE
Semantics with Conjunctions & Embedding

### DISCUSSION THREAD

**Discussion by verdverm** [cue-lang/cue#1090] [Q&A] [answered]
I'm wondering if the following is working as intended or if there is a bug.

Generally, the idea is to have an open `#Base` definition for some common constraints, add in some `#Extra` fields via embedding, and then have a closed `#Schema` used for validation. The embedding seems to negate the conjunction in `Schema2` where I expected `name` to still be a valid and required (for concreteness) field.

```
#Base: {
        name: string
        ...
}

#Extra: {
        foo: string
}

#Schema1: #Base & {
        kind: string
}

#Schema2: #Base & {
        kind: string
        #Extra
}

v1: #Schema1 & {
        name: "v1"
        kind: "v1"
}

v2: #Schema2 & {
        name: "v2"
        kind: "v2"
}
```

`cue eval` reports the following

```
#Schema2: field `name` not allowed:
    ./building-up-repro.cue:2:2
    ./building-up-repro.cue:6:9
    ./building-up-repro.cue:14:11
    ./building-up-repro.cue:14:19
    ./building-up-repro.cue:16:2
v2: field `name` not allowed:
    ./building-up-repro.cue:2:2
    ./building-up-repro.cue:6:9
    ./building-up-repro.cue:14:11
    ./building-up-repro.cue:14:19
    ./building-up-repro.cue:16:2
    ./building-up-repro.cue:24:5
    ./building-up-repro.cue:25:2
 ```
 
 Note, this did work in the v0.3.0 { late alpha, early beta } versions, though that's not saying much ;]

**Comment by lks128** [comment, marked as answer]
This is working as intended because:

> Structs defined by definitions are implicitly closed. ([Definitions](https://cuelang.org/docs/tutorials/tour/types/defs/))

**Reply by verdverm** [reply]
Except that you can tell CUE to leave a definition open with `...`, note that `#Schema1` is valid above

https://cuelang.org/docs/tutorials/tour/intro/schema/

**Reply by lks128** [reply]
Thanks, definitely missed that point.

I noticed that including `...` into `#Extra` makes it work. Looks like it first merges `#Base & #Extra` which gives us a closed struct before it gets a chance to merge with `{ name: "v2", kind: "v2" }`.

**Reply by lks128** [reply]
One question though, is it expected that `#Schema` is an open struct?
```
#Base: {
        name: string
        ...
}

#Schema1: #Base & {
        kind: string
}

v1: #Schema1 & {
        name: "v1"
        kind: "v1"
        abc: 1
}
```
Produces with `v0.4.0`: 
```
v1: {
	name: "v1"
	kind: "v1"
	abc:  1
}
```

**Reply by verdverm** [reply]
hmm, probably as it is a conjunction with an open struct and thus probably remains open?

**Reply by verdverm** [reply]
Regardless, `name` is in the base and is not allowed only under the embedded case

**Reply by myitcv** [reply]
This appears to be working as intended. 

In:

```
#Schema2: #Base & {
	kind: string
	#Extra
}
```

the RHS is a conjunction between `#Base` and:

```
{
	kind: string
	#Extra
}
```

The second conjunct is closed as a result of the embedding of the definition `#Extra`, and because that definition is not open, the field `name` is correctly flagged as not being permitted.

@verdverm perhaps you intended the following?

```
#Schema2: {
    #Base 
    #Extra
    kind: string
}
```

(which also obviates the need for `#Base` to be left open, because embedding disregards closedness per https://tip.cuelang.org/docs/references/spec/#embedding)

**Reply by verdverm** [reply]
Ah, RHS... LHS... I see the algebra now, thanks!

**Reply by myitcv** [reply]
@verdverm - marked this thread as the answer for those following along - let me know if you disagree?

**Reply by verdverm** [reply]
Thinking about it some more, given this explanation, why is it that `#Schema1` works while 2 does not?

What is it about the embedding that changes the semantics?

My expectation was that in the following, the two S's would be equivalent

```
#Extra: foo: string

#S1: {
  bar: string
  foo: string
}

#S2: {
  bar: string
  #Extra
}
```

**Reply by myitcv** [reply]
From the spec:

> A struct may contain an embedded value, an operand used as a declaration. An embedded value of type struct is unified with the struct in which it is embedded, but _disregarding the restrictions imposed by closed structs_. So if an embedding resolves to a closed struct, the corresponding enclosing struct will also be closed, but may have fields that are not allowed if normal rules for closed structs were observed.

(emphasis my own).

**Reply by verdverm** [reply]
That doesn't explain why `Schema1` works and `Schema2` does not. I would say this question remains unanswered to me

**Reply by myitcv** [reply]
Ah sorry, I thought this was a reference to the embedding version. I was lazy and didn't check above.

By `#Schema1` I assume we're referring to:

```
#Base: {
        name: string
        ...
}

#Extra: {
        foo: string
}

#Schema1: #Base & {
        kind: string
}

v1: #Schema1 & {
        name: "v1"
        kind: "v1"
}
```

If so, what in particular is not clear? `#Schema1` is a conjunction of an open definition (`#Base`) and a non-closed struct (`{kind: string}`), the result of which is an open struct. Hence the unification to form `v1` succeeds.

**Reply by jpluscplusm** [reply]
@myitcv I've just arrived at this discussion while trying to work out from first principles the closed->closed->open or open->open->closed hierarchy of schema design that's most aligned with how cue works.

Modifying these examples slightly, it's not clear to me (cue v0.4) why `Concrete1`/`Concrete2` are allowed to extend `#Extra`, and add `why_does_this_not_error`:

```
#Base: {
        name: string
        ...
}

#Extra: {
	#Base
        foo: string
}

Concrete1: #Extra
Concrete1: {
	name: "a-name"
	foo:  "a-string"
	why_does_this_not_error: true
}

Concrete2: #Extra & {
	name: "a-name"
	foo:  "a-string"
	why_does_this_not_error: true
}
```

`cue vet` has no problems with this.

I don't quite grok how to re-close a previously-opened struct/definition, and define a closed type that can be robustly validated against.

In my example, above, how do I define `#Extra` as being something which extends `#Base`, but *is now closed*, so that either `Concrete` variant can't add fields?

I /thought/ that deliberately specifying `#Extra` as a *definition*, without a trailing `...`, should achieve that but, as per my example, it doesn't. `ConcreteX` can do whatever they like, it appears ... :-)

**Reply by verdverm** [reply]
I suspect the ellipses come along for the ride when `#Base` is embedded in the new definition

I thought that using `close()` might help, but it does not seem  to 

```
#Extra: close({
  #Base
  foo: string
})
```

I am unclear what the expected and desired behavior might be with this embedded open definition example

**Reply by myitcv** [reply]
@jpluscplusm your example is working as expected. The `...` is leaving `#Extra` open. The answer here is to not include `...` in `#Base` - it is superfluous, because the very act of embedding `#Base` in `#Extra` allows you to define additional fields in `#Extra`. That was the reason for including my quote from the spec in https://github.com/cue-lang/cue/discussions/1090#discussioncomment-1032784

**Reply by myitcv** [reply]
```
exec cue vet x.cue

-- x.cue --
#Base: {
	name: string
}

#Extra: {
	#Base
	foo: string
}

Concrete1: #Extra
Concrete1: {
	name:                    "a-name"
	foo:                     "a-string"
	why_does_this_not_error: true
}

Concrete2: #Extra & {
	name:                    "a-name"
	foo:                     "a-string"
	why_does_this_not_error: true
}
```

behaves as expected:

```
> exec cue vet x.cue
[stderr]
Concrete1: field not allowed: why_does_this_not_error:
    ./x.cue:1:8
    ./x.cue:5:9
    ./x.cue:6:2
    ./x.cue:10:12
    ./x.cue:14:2
Concrete2: field not allowed: why_does_this_not_error:
    ./x.cue:1:8
    ./x.cue:5:9
    ./x.cue:6:2
    ./x.cue:17:12
    ./x.cue:20:2
[exit status 1]
FAIL: /tmp/testscript827568555/repro.txt/script.txt:1: unexpected command failure
```
